pkg/cache/cache.go

```diff
@@ -0,0 +1,195 @@
+package cache
+
+import (
+	"fmt"
+	"github.com/patrickmn/go-cache"
+	"k8s.io/client-go/util/workqueue"
+	//"reflect"
+	glog "github.com/Sirupsen/logrus"
+	"k8s.io/client-go/pkg/api/v1"
+	k8sCache "k8s.io/client-go/tools/cache"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	calicoClient "github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/api/unversioned"
+	"github.com/kylelemons/godebug/pretty"
+)
+
+type ResourceCache interface {
+	// Sets the key to the provided value, and generates an update
+	// on the queue the value has changed.
+	Set(key string, value interface{})
+
+	// Gets the value associated with the given key.  Returns nil
+	// if the key is not present.
+	Get(key string) (interface{}, bool)
+
+	// Sets the key to the provided value, but does not generate
+	// and update on the queue ever.
+	// Priming should happen only at the begining and only run 
+	// method should have access to it. Hence a private method.
+	prime(key string, value interface{})
+
+	// Deletes the value identified by the given key from the cache, and
+	// generates an update on the queue if a value was deleted.
+	Delete(key string)
+
+	// Lists the keys currently in the cache.
+	ListKeys() []string
+
+	// Returns true when the cache has synced with the datastore,
+	// false until that point.
+	HasSynced() bool
+
+	// Starts the cache.
+	Run(stopChan chan struct{})
+
+	// Sets kubernetes cache indexer
+	SetIndexer(indexer k8sCache.Indexer)
+}
+
+// CalicoCache implements ResourceCache interface
+type CalicoCache struct {
+	threadSafeCache *cache.Cache
+	workqueue       workqueue.RateLimitingInterface
+	calicoClient    *calicoClient.Client
+	synced          bool 							// Flag set to true when cache is synced with etcd datastore
+	objType			string							// Type of object CalicoCache will hold. E.g. calico profiles, policies etc.
+	k8sCacheIndexer k8sCache.Indexer				// K8S cache indexer used while priming of calicoCache.
+}
+
+func NewCache(queue workqueue.RateLimitingInterface, client *calicoClient.Client, objectType string) ResourceCache {
+	
+	return &CalicoCache{
+		threadSafeCache: cache.New(cache.NoExpiration, cache.DefaultExpiration),
+		workqueue:       queue,
+		calicoClient:    client,
+		synced:			 false,
+		objType:		 objectType,
+	}
+}
+
+func (c *CalicoCache) Set(key string, obj interface{}) {
+
+	
+	// Converting obj to calico representation
+	profile := api.Profile{
```

Comment:
I'm not sure "primer" and "reconciler" are the right abstractions - I was thinking something like this: https://gist.github.com/caseydavenport/53421b1407bbb5059e13d13a2ec1ec57#file-resource-cache-interface-L40

- A `ListFunc` which lists all the relevant objects from etcd
- A `KeyFunc` which generates a string key from one of those objects.

The reconciling / priming logic is, I think, the same across all resource types, so it would be nice to hide that within the cache rather than behind a a Primer / Reconciler interface that needs to be implemented for each resource.  Does that make sense?